platform/mbed_wait_api_rtos.cpp

```diff
@@ -23,32 +23,55 @@
 #include "rtos/rtos.h"
 #include "platform/mbed_critical.h"
 #include "platform/mbed_power_mgmt.h"
+#include "platform/mbed_error.h"
 
 void wait(float s)
 {
-    wait_us(s * 1000000.0f);
-}
-
-void wait_ms(int ms)
-{
-    wait_us(ms * 1000);
-}
+    if ((s >= 0.01f)  && core_util_are_interrupts_enabled()) {
+        wait_ms(s * 1000.0f);
+        return;
+    }
 
-void wait_us(int us)
-{
+    uint32_t us = (s * 1000000.0f);
     const ticker_data_t *const ticker = get_us_ticker_data();
-
     uint32_t start = ticker_read(ticker);
     if ((us >= 1000) && core_util_are_interrupts_enabled()) {
         // Use the RTOS to wait for millisecond delays if possible
         sleep_manager_lock_deep_sleep();
-        Thread::wait((uint32_t)us / 1000);
+        ThisThread::sleep_for((uint32_t)us / 1000);
         sleep_manager_unlock_deep_sleep();
     }
     // Use busy waiting for sub-millisecond delays, or for the whole
     // interval if interrupts are not enabled
     while ((ticker_read(ticker) - start) < (uint32_t)us);
 }
 
+/*  The actual time delay may be up to one timer tick less - 1 msec */
+void wait_ms(int ms)
+{
+    if (core_util_is_isr_active() || !core_util_are_interrupts_enabled()) {
+#if defined(MBED_TRAP_ERRORS_ENABLED) && MBED_TRAP_ERRORS_ENABLED
+        MBED_ERROR(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_INVALID_OPERATION),
+                    "Deprecated behavior: milli-sec delay should not be used in interrupt.\n");
+#else
+        wait_us(ms * 1000);
```

Comment:
As stated by Kevin, we could use the low power ticker in that case. 
It would be useful to at least print a warning here as behaviour with and without `MBED_TRAP_ERRORS_ENABLED` sets are too very different (trap or execute the intended action). 
Note that this is a breaking change and it should be documented as one.

Reply:
Ok we either trap or not, having it error out in one mode, but happily work without even a warning, in other doesn't make much sense to me.

I don't get this part and I really hope every use of `MBED_ERROR` doesn't forces users to add this `if`:
```
#if defined(MBED_TRAP_ERRORS_ENABLED) && MBED_TRAP_ERRORS_ENABLED
        MBED_ERROR(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_INVALID_OPERATION),
                    "Deprecated behavior: milli-sec delay should not be used in interrupt.\n");
#else
```

I'm happy to show a warning and change the behaviour for next major.